### PR TITLE
Refactor the build scripts

### PR DESCRIPTION
**Pull request by gaearon**
This might be a bit hard to review since a few things moved around.
The code that produces the bundles is exactly the same (just split into a few methods).

This mostly changes the code that copies folders and shims:

* We no longer `process.exit()` all over the place, and instead use the normal try/catch flow.
* We intentionally crash on uncaught Promise rejections.
* The old code was convoluted because it tried to create an npm package folder lazily for every bundle. The new code just creates them at the end (so we don't need to check if one exists).
* Another reason the old code was convoluted is due to all the checks if folders exist before creating them. I just changed our `asyncCopyTo` function to do that automatically.
* I'm no longer creating npm packages in a temp folder, and instead I'm doing it in place. This is easier to read and understand. We're also now doing one pack/unpack cycle per package (instead of per bundle).

Note I mostly rewrote the folder moving file from scratch.
I found it easier than refactoring the old one.

### CODE REVIEW

File: scripts/rollup/build.js [hunk 1]
```diff
@@ -10,33 +10,40 @@ const stripBanner = require('rollup-plugin-strip-banner');
...
-const syncReactNative = require('./sync').syncReactNative;
-const syncReactNativeRT = require('./sync').syncReactNativeRT;
-const syncReactNativeCS = require('./sync').syncReactNativeCS;
+const {
+  syncReactDom,
+  syncReactNative,
+  syncReactNativeRT,
+  syncReactNativeCS,
+} = require('./sync').syncReactDom;
 const Packaging = require('./packaging');
```
Comment (trueadm) [inline scripts/rollup/build.js]: Can we also deconstruct this object rather than using `Packaging`?

**Comment by gaearon** [comment]
I wanted to keep `Modules` and `Packages` the way they are so it's obvious these aren't local functions. I think this helps understand the structure a little bit better.

I'll do the same for `Sync` instead to be consistent.

File: scripts/rollup/build.js [hunk 2]
```diff
@@ -365,139 +332,154 @@ async function createBundle(bundle, bundleType) {
...
-      npmPackagesTmpDir
-    );
-    console.log(`${chalk.bgGreen.black(' COMPLETE ')} ${logKey}\n`);
+    const result = await rollup(rollupConfig);
+    await result.write(rollupOutputOptions);
   } catch (error) {
-    if (error.code) {
-      console.error(
-        `\x1b[31m-- ${error.code}${error.plugin ? ` (${error.plugin})` : ''} --`
+    console.log(`${chalk.bgRed.black(' OH NOES! ')} ${logKey}\n`);
```
Comment (bvaughn) [inline scripts/rollup/build.js]: nit: `console.error` ?

File: scripts/rollup/build.js [hunk 3]
```diff
@@ -365,139 +332,154 @@ async function createBundle(bundle, bundleType) {
...
-    // create the dist folder for UMD bundles
-    fs.mkdirSync(path.join('build', 'dist'));
+function handleRollupError(error) {
+  loggedErrors.add(error);
+  if (!error.code) {
+    console.error(error);
+    return;
+  }
+  console.error(
+    `\x1b[31m-- ${error.code}${error.plugin ? ` (${error.plugin})` : ''} --`
```
Comment (bvaughn) [inline scripts/rollup/build.js]: Why not use `chalk.red()` ?

File: scripts/rollup/utils.js [hunk 4]
```diff
@@ -1,60 +1,89 @@
...
-
-function resolvePath(path) {
-  if (path[0] === '~') {
-    return join(process.env.HOME, path.slice(1));
-  } else {
-    return resolve(path);
-  }
+  return asyncMkDirP(path.dirname(to)).then(
+    () =>
+      new Promise((resolve, reject) => {
```
Comment (bvaughn) [inline scripts/rollup/utils.js]: Possible follow-up: We could remove this `new Promise` boilerplate here and below using [`Promise.denodeify`](https://www.npmjs.com/package/promise#promisedenodeifyfn)

File: scripts/rollup/build.js (hunk 3 above)
Comment (gaearon) [reply scripts/rollup/build.js]: Copy paste of existing code, didn't look into how it works.

File: scripts/rollup/build.js (hunk 2 above)
Comment (gaearon) [reply scripts/rollup/build.js]: There's a further console.error too so I guess it doesn't matter? The diff makes it look like I replaced one with the other, but I just added an additional one.

Comment (gaearon) [reply scripts/rollup/build.js]: To be honest I'm not even sure what the difference on Node is 😅

File: scripts/rollup/build.js (hunk 3 above)
Comment (bvaughn) [reply scripts/rollup/build.js]: Gotcha. Should be identical to `chalk.red('string')`

File: scripts/rollup/build.js (hunk 2 above)
Comment (bvaughn) [reply scripts/rollup/build.js]: I think the only difference is that one outputs to `stderr` and the other to `stdout`. I don't think it actually matters in this case. 😁

**Pull request merged**